### PR TITLE
Starting using DocumentationAnalyzers

### LINKED ISSUE #21094: Start using Roslyn documentation analyzers
https://github.com/DotNetAnalyzers/DocumentationAnalyzers

Suggested by @sharwell in https://github.com/dotnet/efcore/issues/20849#issuecomment-636528238

### PR DESCRIPTION
**Pull request by roji**
Just racking up the consistency points on a fine Monday morning.

Closes #21094

### CODE REVIEW

File: src/Microsoft.Data.Sqlite.Core/SqliteTransaction.cs [hunk 1]
```diff
@@ -117,7 +117,7 @@ public override void Rollback()
         ///     Releases any resources used by the transaction and rolls it back.
         /// </summary>
         /// <param name="disposing">
-        ///     <see langword="talse"/> to release managed and unmanaged resources;
+        ///     <see langword="true"/> to release managed and unmanaged resources;
```
Comment (roji) [inline src/Microsoft.Data.Sqlite.Core/SqliteTransaction.cs:120]: talse!

File: EFCore.ruleset [hunk 2]
```diff
@@ -165,4 +165,22 @@
     <Rule Id="RCS1194" Action="None" />
     <Rule Id="RCS1206" Action="None" />
   </Rules>
+  <Rules AnalyzerId="DocumentationAnalyzers" RuleNamespace="DocumentationAnalyzers">
+    <Rule Id="DOC100" Action="Warning" /> <!-- PlaceTextInParagraphs -->
```
Comment (sharwell) [inline EFCore.ruleset:169]: 📝 This one is often set to `Hidden` or `None`. It probably isn't necessary unless you are using Sandcastle Help File Builder.

File: src/EFCore/Extensions/EntityTypeExtensions.cs [hunk 3]
```diff
@@ -410,9 +410,10 @@ public static bool IsOwned([NotNull] this IEntityType entityType)
         ///     Gets a value indicating whether given entity type is in ownership path for this entity type.
         /// </summary>
         /// <param name="entityType"> The entity type. </param>
-        /// <param name="targetType"> Entity type to search for in ownership path. </param>
+        /// <param name="targetType"> Entity type to search for in ownership path.
         ///     <see langword="true" /> if <paramref name="targetType" /> is in ownership path of <paramref name="entityType" />,
         ///     otherwise <see langword="false" />.
+        ///  </param>
```
Comment (sharwell) [inline src/EFCore/Extensions/EntityTypeExtensions.cs:416]: 💭 It seems the original `param` element was correct, and the missing item is the `<returns>` element around the final two lines.

Comment (roji) [reply src/EFCore/Extensions/EntityTypeExtensions.cs:416]: You're probably right, I'll fix this - thanks.

File: EFCore.ruleset (hunk 2 above)
Comment (roji) [reply EFCore.ruleset]: Makes sense to me, reverted these changes.

**Comment by bricelam** [comment]
Should we dump StyleCop? (We only use it for its documentation rules)

**Comment by sharwell** [comment]
> Should we dump StyleCop? (We only use it for its documentation rules)

There isn't a bunch of overlap between the two analyzer sets. Many of the rules in DocumentationAnalyzers were added after they were rejected from StyleCop Analyzers (most often for being too vague or too specific to certain developer scenarios). I'd need to see more details about which rules are used to give a better answer.

**Comment by roji** [comment]
@bricelam aren't there other, non-docs-related rules we could enable from StyleCop? Or is FxCop the one we want (#18870)?

**Comment by sharwell** [comment]
For code style rules, the in-box experience from .editorconfig is intended to provide the most seamless experience. We're currently in the process of making a separate analyzer package for command line enforcement of these rules. Other analyzer packages generally have varying degrees of overlap while often offering unique features as well. Each analyzer incurs a small cost, so the team would need to decide which ones are providing the most overall value.

File: src/EFCore/Diagnostics/WarningsConfiguration.cs [hunk 4]
```diff
@@ -134,8 +134,10 @@ [NotNull] IEnumerable<(EventId Id, LogLevel Level)> eventsAndLevels)
                 : null;
 
         /// <summary>
-        ///     Gets the <see cref="LogLevel" /> set for the given event ID, or <code>null</code>
-        ///     if no explicit behavior has been set.
+        ///     <para>
+        ///         Gets the <see cref="LogLevel" /> set for the given event ID, or <see langword="null"/>
+        ///         if no explicit behavior has been set.
+        ///     </para>
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/WarningsConfiguration.cs:140]: This `para` is redundant

Comment (AndriySvyryd) [reply src/EFCore/Diagnostics/WarningsConfiguration.cs:140]: Missing `returns`

File: src/Directory.Build.props [hunk 5]
```diff
@@ -1,8 +1,12 @@
...
   <Import Project="..\Directory.Build.props" />
-  
+
   <PropertyGroup>
     <IsPackable>True</IsPackable>
     <IncludeSymbols>True</IncludeSymbols>
   </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="DotNetAnalyzers.DocumentationAnalyzers" Version="$(DocumentationAnalyzersVersion)" PrivateAssets="All" />
```
Comment (smitpatel) [inline src/Directory.Build.props:10]: I think according to naming convention this should be `DotNetAnalyzersDocumentationAnalyzersVersion`

File: src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs [hunk 6]
```diff
@@ -20,8 +20,10 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Conventions
...
     ///         * [entity name]Id
     ///     </para>
-    ///     If the entity type is owned through a reference navigation property then the corresponding foreign key
-    ///     properties are used.
+    ///     <para>
+    ///         If the entity type is owned through a reference navigation property then the corresponding foreign key
+    ///         properties are used.
+    ///     </para>
     ///     <para>
     ///     </para>
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:28]: empty `para`

File: src/Microsoft.Data.Sqlite.Core/SqliteTransaction.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/Microsoft.Data.Sqlite.Core/SqliteTransaction.cs:120]: What begins with "t" and is not false?

Comment (roji) [reply src/Microsoft.Data.Sqlite.Core/SqliteTransaction.cs:120]: That must be the 3-value logic thing everyone's been talking about...

File: src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs (hunk 6 above)
Comment (roji) [reply src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:28]: Weird that this wasn't flagged, /cc @sharwell

Comment (sharwell) [reply src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs]: I would recommend filing bug reports over on the analyzer project for cases like this.

**Comment by bricelam** [comment]
> @bricelam aren't there other, non-docs-related rules we could enable from StyleCop? Or is FxCop the one we want (#18870)?

No other rules--we just use it to check the correctness of XML docs (e.g. no `<returns>` on a `void` method).

Smit and I think Andriy maintain the .editorconfig file for enforcing our style guidelines. Personally, I'm against enforcing style at all since it limits artistic expression, but on the other hand, I agree that some level of consistency is needed.

I definitely still want to enable more FxCop rules. I'm sure they'll catch some globalization bugs we've missed.

**Comment by AndriySvyryd** [comment]
> it limits artistic expression

One of the most important aspects in software engineering 😆

**Pull request merged**